### PR TITLE
Basic ADS client

### PR DESCRIPTION
**Pull request by renuka-fernando**
Add basic ADS client

Related issue: https://github.com/envoyproxy/ratelimit/issues/201
Related PR: https://github.com/envoyproxy/ratelimit/pull/373

Signed-off-by: Renuka Fernando <renukapiyumal@gmail.com>

### CODE REVIEW

**Comment by renuka-fernando** [comment]
Hey @alecholmez,

I found this xDS client sample, https://github.com/grpc/grpc-go/tree/master/examples/features/xds
But I am not sure that we can use this, in its sample client, it is just a call to the method "SayHello". It is not to send the Discovery Request.

Client: https://github.com/grpc/grpc-go/blob/master/examples/features/xds/client/main.go#L68

**Comment by alecholmez** [comment]
So we can actually do something similar, I've written an envoy xDS client before let me see if I can dig that up...

This code: https://github.com/greymatter-io/gm-fabric-go/blob/main/discovery/xds.go
Is rather old and only works for V2 and hasn't been touched in years haha, but it's a similar concept.

**Comment by renuka-fernando** [comment]
Thank you for your response. But the link https://github.com/greymatter-io/gm-fabric-go/blob/main/discovery/xds.go is not visible to me, it gives me 404.

**Comment by renuka-fernando** [comment]
Hey, @alecholmez,

When you get a chance, could you please review this PR?

**Comment by renuka-fernando** [comment]
Hi @kyessenov @sunjayBhatia @alecholmez ,

Could you please have a look? This is related to the envoy rate limit service configuration update feature via xDS protocol. There, rate limit service works as the xDS client and GCP as xDS server. Since the rate limit service requires an xDS client in golang we wanted to have a generic golang xDS client in GCP repo. It is really grateful if you could have a look.

Thanks.

**Comment by renuka-fernando** [comment]
Hi @alecholmez,

Could you please have a look? Please let me know if any changes are required for the basic ADS client.

File: pkg/adsclient/sotw/v3/client.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/golang/protobuf/ptypes/any"
+	status "google.golang.org/genproto/googleapis/rpc/status"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	grpcStatus "google.golang.org/grpc/status"
+)
+
+var (
+	errInit    = errors.New("ads client: grpc connection is not initialized (use InitConnect() method to initialize connection)")
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:34]: Can we export these? Having users interact with errors using
```go
errors.Is(err, ErrInit)
```
is a pleasant user experience IMO.

File: pkg/adsclient/sotw/v3/client.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	errNilResp = errors.New("ads client: nil response from xds management server")
+)
+
+type AdsClient interface {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:38]: Can you write a comment on this about what the interface does and why it exists?

I'd also recommend changing the name to:
```go
type ADSClient interface {
```

File: pkg/adsclient/sotw/v3/client.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	InitConnect(clientConn grpc.ClientConnInterface, opts ...grpc.CallOption) error
+	// Fetch waits for a response from management server and returns response or error.
+	Fetch() (*Response, error)
+	// Ack acknowledge the validity of the last received response to management server.
+	Ack() error
+	// Nack acknowledge the invalidity of the last received response to management server.
+	Nack(message string) error
+}
+
+type Response struct {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:49]: Can you write a comment why we need this wrapper?

File: pkg/adsclient/sotw/v3/client.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+	// streamClient is the ADS discovery client
+	streamClient discovery.AggregatedDiscoveryService_StreamAggregatedResourcesClient
+	// lastAckedResponse is the last response acked by the ADS client
+	lastAckedResponse *discovery.DiscoveryResponse
+	// lastReceivedResponse is the last response received from management server
+	lastReceivedResponse *discovery.DiscoveryResponse
+}
+
+func NewAdsClient(ctx context.Context, nodeID string, typeURL string) AdsClient {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:67]:
```suggestion
func NewADSClient(ctx context.Context, nodeID string, typeURL string) ADSClient {
```

File: pkg/adsclient/sotw/v3/client.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	return &adsClient{
+		ctx:     ctx,
+		nodeID:  nodeID,
+		typeURL: typeURL,
+	}
+}
+
+func (c *adsClient) InitConnect(clientConn grpc.ClientConnInterface, opts ...grpc.CallOption) error {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:75]: Can you add doc strings to the exported methods on this struct?

File: pkg/adsclient/sotw/v3/client.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+	streamClient, err := discovery.NewAggregatedDiscoveryServiceClient(clientConn).StreamAggregatedResources(c.ctx, opts...)
+	if err != nil {
+		return err
+	}
+	c.streamClient = streamClient
+	return c.Ack()
+}
+
+func (c *adsClient) Fetch() (*Response, error) {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:84]: Doc string

File: pkg/adsclient/sotw/v3/client.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+	c.mu.Lock()
+	c.lastReceivedResponse = resp
+	c.mu.Unlock()
+
+	return &Response{
+		Resources: resp.GetResources(),
+	}, err
+}
+
+func (c *adsClient) Ack() error {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:105]: Doc string

File: pkg/adsclient/sotw/v3/client.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+	c.mu.Lock()
+	c.lastAckedResponse = c.lastReceivedResponse
+	c.mu.Unlock()
+	return c.send(nil)
+}
+
+func (c *adsClient) Nack(message string) error {
```
Comment (alecholmez) [inline pkg/adsclient/sotw/v3/client.go:112]: Docstring

**Review by alecholmez — CHANGES_REQUESTED** [review]
I think overall this looks good. Just some cleanup here and there and we can get this in.

One recommendation I might have would be to change the folder structure:
```
pkg/client/sotw/v3/client.go
```
I imagine we might want more of these. Maybe a delta client, xDS client, delta xDS client, etc...

File: pkg/adsclient/sotw/v3/client.go (hunk 2 above)
Comment (renuka-fernando) [reply pkg/adsclient/sotw/v3/client.go]: Fixed with https://github.com/envoyproxy/go-control-plane/commit/1ab595617393939429f91b8edaa479194b9c7654

File: pkg/adsclient/sotw/v3/client.go (hunk 3 above)
Comment (renuka-fernando) [reply pkg/adsclient/sotw/v3/client.go]: Added with https://github.com/envoyproxy/go-control-plane/commit/1ab595617393939429f91b8edaa479194b9c7654

File: pkg/adsclient/sotw/v3/client.go (hunk 4 above)
Comment (renuka-fernando) [reply pkg/adsclient/sotw/v3/client.go]: Fixed with https://github.com/envoyproxy/go-control-plane/commit/1ab595617393939429f91b8edaa479194b9c7654

File: pkg/adsclient/sotw/v3/client.go (hunk 5 above)
Comment (renuka-fernando) [reply pkg/adsclient/sotw/v3/client.go]: Added with https://github.com/envoyproxy/go-control-plane/commit/1ab595617393939429f91b8edaa479194b9c7654

**Comment by renuka-fernando** [comment]
> I think overall this looks good. Just some cleanup here and there and we can get this in.
> 
> One recommendation I might have would be to change the folder structure:
> 
> ```
> pkg/client/sotw/v3/client.go
> ```
> 
> I imagine we might want more of these. Maybe a delta client, xDS client, delta xDS client, etc...

Hi @alecholmez,

Thank you for your comments.

Moved the client.go file as requested to `pkg/client/sotw/v3/client.go`.
So later we can add delta client as `pkg/client/delta/v3/client.go`.

**Comment by alecholmez** [comment]
Great! I'm letting the build run and hopefully we can get this in shortly.

**Comment by alecholmez** [comment]
@renuka-fernando looks like the build failed. Can you look into this?

**Comment by renuka-fernando** [comment]
@alecholmez could you please check now?

**Comment by renuka-fernando** [comment]
Hi @alecholmez,

Could you please run the failed build again and please review the PR? The issue with the build is fixed now: https://github.com/envoyproxy/go-control-plane/issues/625

**Comment by alecholmez** [comment]
@renuka-fernando looks like this one still failed can you check it out?

**Review by alecholmez — APPROVED** [review]
This looks good thanks for the contributions!

**Pull request merged**